error: cannot convert 'struct array' to 'short'

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Pop!_OS 20.04 LTS
Processor: 24 cpus, 64bit, little endian, AMD Ryzen 9 3900 12-Core Processor
CC version: cc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0

getwd: /home/refaqtor/Sync/Work/SeaArctosLLC/data/sqliter
vmodules: /home/refaqtor/.vmodules
vroot: /home/refaqtor/TOOLS/v
vexe: /home/refaqtor/TOOLS/v/v
vexe mtime: 2021-09-12 19:50:41
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 c847152.cd7d482

Git version: git version 2.23.0
Git vroot status: weekly.2021.36-46-gcd7d482c
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg cannotConvertCompilerBug.v`
```v
import encoding.hex as hx

bytes := '010000150107120508d37445a0d7e5c5071980710c64310d9e12043000777369ff0424ab78b91a05164b00e50034003300'
yr := i16((hx.decode(bytes.substr(6,8))?))
println(bytes)
println('yr: $yr')

```

**What did you expect to see?**

successful compile, or at least, not 'This is a compiler bug' message


**What did you see instead?**
```
==================
/tmp/v/cannotConvertCompilerBug.16348920628964752191.tmp.c:10489: error: cannot convert 'struct array' to 'short'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2506,6 +2506,10 @@ pub fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		snexpr := node.expr.str()
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast string to `$tt`, use `${snexpr}.str` instead.', node.pos)
+	} else if final_from_sym.kind == .array && !from_type.is_ptr() && to_type != ast.string_type {
+		ft := c.table.type_to_str(from_type)
+		tt := c.table.type_to_str(to_type)
+		c.error('cannot cast array `$ft` to `$tt`', node.pos)
 	}
 
 	if to_sym.kind == .rune && from_sym.is_string() {

--- a/vlib/v/checker/tests/cast_array_to_number_err.out
+++ b/vlib/v/checker/tests/cast_array_to_number_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/cast_array_to_number_err.vv:5:8: error: cannot cast array `[]u8` to `i16`
+    3 | fn main() {
+    4 |     bytes := '010000150107120508d37445a0d7e5c5071980710c64310d9e12043000777369ff0424ab78b91a05164b00e50034003300'
+    5 |     yr := i16(hx.decode(bytes.substr(6,8))?)
+      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    6 |     println(bytes)
+    7 |     println('yr: $yr')

--- a/vlib/v/checker/tests/cast_array_to_number_err.vv
+++ b/vlib/v/checker/tests/cast_array_to_number_err.vv
@@ -1,0 +1,8 @@
+import encoding.hex as hx
+
+fn main() {
+	bytes := '010000150107120508d37445a0d7e5c5071980710c64310d9e12043000777369ff0424ab78b91a05164b00e50034003300'
+	yr := i16(hx.decode(bytes.substr(6,8))?)
+	println(bytes)
+	println('yr: $yr')
+}

